debezium-core/src/main/java/io/debezium/pipeline/signal/channels/process/InProcessSignalChannel.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.pipeline.signal.channels.process;
+
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.DebeziumException;
+import io.debezium.annotation.ThreadSafe;
+import io.debezium.config.CommonConnectorConfig;
+import io.debezium.pipeline.signal.SignalRecord;
+import io.debezium.pipeline.signal.channels.SignalChannelReader;
+
+/**
+ * Implementation of {@link SignalChannelReader} that also implements {@link SignalChannelWriter}
+ * used for sending signal from the same JVM process
+ *
+ * <p>
+ *     Mainly targeted at Debezium Engine
+ * </p>
+ */
+@ThreadSafe
+public class InProcessSignalChannel implements SignalChannelReader, SignalChannelWriter {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InProcessSignalChannel.class);
+    public static final String CHANNEL_NAME = "in-process";
+
+    private final AtomicBoolean open = new AtomicBoolean(false);
+    private final Queue<SignalRecord> signals = new ConcurrentLinkedQueue<>();
+
+    @Override
+    public String name() {
+        return CHANNEL_NAME;
+    }
+
+    @Override
+    public void init(CommonConnectorConfig connectorConfig) {
+        open.compareAndSet(false, true);
+        LOGGER.info("Reading signals from {} channel", CHANNEL_NAME);
+    }
+
+    @Override
+    public List<SignalRecord> read() {
+        if (!open.get()) {
+            return List.of();
+        }
+        return Stream.ofNullable(signals.poll()).toList();
+    }
+
+    @Override
+    public void close() {
+        if (open.compareAndSet(true, false)) {
+            drain();
+        }
+    }
```

Comment:
@vjuranek  @jpechane  This is not sufficient and in theory `close()` can start draining  after concurrent call to `read` already passed the `open.get()` check. So a lock would be in order to properly prevent this

However again -- I don't think there is a problem if that happens as such order of operation is equal to eider calling `read()` and then `close()` serially or the other way around -- the connector code  operating the channel must be able to handle any of these orders and thus locking here is pointless IMHO.

Reply:
I'd at least add comment here with the reasoning why locking doesn't matter.